### ISSUE TITLE
Instance for Maybe loses information

### ISSUE THREAD

**Issue by cgibbard** [bos/aeson#376]
```
Prelude Data.Aeson> encode (Just Nothing :: Maybe (Maybe Integer))
"null"
Prelude Data.Aeson> encode (Nothing :: Maybe (Maybe Integer))
"null"
```

This just cost me way more time than I'd care to admit. Personally, I think it'd be worth fixing it by using an instance similar to the one used for Either or other sum types. Just letting TH generate the instance would be fine I think. You might want to do a major version increment to prepare people for the fact that the representation is different, but I think it's important not to have default ToJSON/FromJSON instances losing information like this.

**Comment by bergmark** [comment]
This instance has been in aeson for over 4 years (since 0.6..0.0). Changing it would silently break all usages and require a new type to restore the old behavior.

I think a first step towards pushing this change through can be to release a library that introduces a new type with an appropriate instance. If it becomes popular we could move towards a migration plan for changing the instance.

**Comment by mgsloan** [comment]
It is possible to write an instance for `Maybe` which ensures its argument is not `Maybe`: https://github.com/ghcjs/ghcjs-base/pull/34/commits/6bf43075849fdd551f04a5b9d8a36e4d769e6fc2#diff-e09b67a14d374b52d6408c4ebcb9b630R7 .  There, I'm using closed type families, but I've just realized this is better:

```haskell
class FromJSON a where
    type FromJSONUsesNull = 'False
    ...

class ToJSON a where
    type ToJSONUsesNull = 'False
    ...
```

Unfortunately, type family defaults aren't usable unless they are associated, so both are needed.  If we want to avoid `DataKinds` (dunno why that would be), we can instead use `ConstraintKinds` and instead have `ToJSONDoesn'tUseNull :: * -> Constraint`.  If it does use null, it'd yield an unsatisfiable constraint like `Error` from the singletons package.

So, it may be possible to keep the behavior but statically ensure that.  Yes, this could cause compilation errors, but they could reflect actual potential errors.

**Comment by cgibbard** [comment]
Yeah, making things fail to compile in cases where this loss of information occurs is probably a good first step -- it would have saved me a fair amount of trouble tracking down where some subtle glitchy behaviour at runtime was coming from.

In our case, we tend to use Aeson on both the backend and frontend of our applications and just use TH-generated instances without much thought about the precise representation of everything in terms of JSON, so changing the instances would break none of our code, but I realise a lot of people using it wouldn't have that luxury. I'd sort of come to expect all communication of Haskell data structures between our frontend and backend to be totally transparent, so it took me a while before I even considered that it might be the JSON encoding and decoding being unfaithful.

**Comment by bergmark** [comment]
This sounds like a good approach!

**Comment by tolysz** [comment]
Problem reoccurring all over: https://github.com/bos/aeson/issues/323
(https://github.com/bos/aeson/pull/324) really is the only sane answer :)
Basically whan we have `ommit nothing` `Nothing` -> `Ommited` otherwise  `Nothing` -> `Null`
And it simply works with any user types...

**Comment by bergmark** [comment]
@tolysz No. There is already a suggested solution here. You also never explained why the suggestions in #323 are not sufficient.

**Comment by tolysz** [comment]
Sorry, I answered it there (now). I simply lost entropy to produce more constructive arguments.
But the basic idea is to always omit the new constructor and always map Null -> `null`. This solution will compose nicely in generic way.

**Comment by tolysz** [comment]
With the proposed additional value
```
Nothing -> Ommited
Just Just a -> toJson a
Just Nothing -> Null

But I guess a newtype wrapper would be better (eg. `Possible`)
```

**Comment by ardamose123** [comment]
Maybe you'd think I'm blaming the victim here, wouldn't it be better to encode a `Maybe (Maybe a))` as something like:

    data MyValue a = OK a | NotOK | SomethingMissing

Maybe then, you'd be able to pattern-match and make it easier to deal with than nested `Maybe`s.

**Comment by cgibbard** [comment]
Well, in the original setting where it came up, we didn't explicitly have
Maybe (Maybe a) appearing as such in the user written code.

It simply came up as a consequence of some type family instances which used
Maybe in an entirely appropriate way.

Of course an even simpler thing would also work: data Maybe' a = Nothing' |
Just' a
(And then derive the relevant aeson instances using TH) But there are
advantages to using Prelude types sometimes.

Our actual solution was just to stick an Identity wrapper in the way.

The main thing which would make me want to change this instance isn't that
it's really hard to work around. It's that it is a case which can show up
and quietly lose data without anyone knowing about it.

On Mon, Jan 9, 2017, 08:56 Ariel D. Moya Sequeira <notifications@github.com>
wrote:

Maybe you'd think I'm blaming the victim here, wouldn't it be better to
encode a Maybe (Maybe a)) as something like:

data MyValue a = OK a | NotOK | SomethingMissing

Maybe then, you'd be able to pattern-match and make it easier to deal with
than nested Maybes.

—
You are receiving this because you authored the thread.

Reply to this email directly, view it on GitHub
<https://github.com/bos/aeson/issues/376#issuecomment-271289970>, or mute
the thread

**Comment by tolysz** [comment]
You are right: `Possible` is the exact type you say.

My only problem is: `Maybe` is special in `aeson`.
Records (`Map`s) are removing `Nothing` for all keys if some template settings are set.

**Comment by cgibbard** [comment]
Of course, it's quite possible that my position on this is a little unique,
since both the producer and consumer of all the JSON I'm  generating are
always Haskell programs using Aeson, so practically the only thing I care
about is that the representation is faithful (that what I put in on one end
is what I get out on the other).

**Comment by phadej** [comment]
I kind of like @mgsloan approach. One possibilty is to make `Value` gadt, so we can control whether it has `Null` constructor. And change `ToJSON` to have functional dependency:

```hs
data NotNull

data Value t where
  Bool :: Bool -> Value t
  Null :: Value Null

class ToJSON t a | a -> t where
    toJSON :: a -> Value t

instance ToJSON t a => ToJSON t [a] where
   ...

instance ToJSON NotNull Bool where
    toJSON = Bool

instance ToJSON NotNull a => ToJSON Null (Maybe a) where
    toJSON Nothing = Null
    toJSON (Just v)  = retag (toJSON v)
```

I''m not sure whether this is worth the trouble.

---

Alternatively in encoding, one can use `Fix (Compose ValueF Maybe)` ([`ValueF`](http://hackage.haskell.org/package/aeson-extra-0.4.0.0/docs/Data-Aeson-Extra-Foldable.html#t:ValueF)`
, so on gets `Object :: HashMap Text (Maybe Value)` like structure, then `Nothing` on that level can be omitted.

I don't like having *top-level* `Omitted`, it makes sense only inside `Object` (and `Array`).

---

Another alternative is to have:

Then one could have `class ToJSONField a where toJSONField :: a  -> Maybe Value` and use it for `Map` / record serialisation. It would be cool to have `IntristicSuperClasses` or something, to provide default implementation `ToJSONField = Just . toJSON`.

---

In summary: there are ideas, but one have to try them (and benchmark!). The second option makes possible to build on existing stuff, so one could explore that.

**Comment by phadej** [comment]
Note: remind me, as at work we use own generics for `ToJSON`, so we could use own class all together exploring `ValueF`. I have not-so-nice fix to make `omittableNothing` with generics-sop, and I don't really like it.

**Comment by cgibbard** [comment]
Yeah, just having our own classes as a replacement for Aeson's ToJSON and
FromJSON would work as well, seeing as we're also using Generics or TH to
generate essentially every instance. But if we ever go to the trouble of
doing that, it practically amounts to forking aeson in a way, which would
be a bit of a shame over such a small thing. It also wouldn't do much to
protect people who are unaware of this sneaky case of data loss. Still,
there are some other cases where the instances aren't quite uniform where
I'd prefer something different. I also found myself wanting to sidestep the
instances for Map to effectively encode Maps the same way as lists of
pairs. (We already have our own wrapper around Map to fix the
Monoid/Semigroup instances and so doing that isn't so hard in our case.)

**Comment by phadej** [comment]
FYI: You can encode maps as a list of pairs if you use wrapper key type with `ToJSONKey` functionality (the default instance btw uses `ToJSON`)

I see aeson using Maybe as a control structure. If you want no-data-loss, you'd need newtype around. That's the problem with not obviously better (or general) instance.

Imho it's important that aeson is convenient, therefore "as a control structure" makes sense as a default

**Comment by tolysz** [comment]
Even if I make a `newtype` I will still have to deal with `omittableNothing` i.e. I will have to modify the type which imports my new one. Either way this is a bit accademic at this point. However I have a working (0.8-is) implementation with a ritcher internal type where all becomes simple.
I will not make a promise, but I can estimate it is about 2-3 evenings of porting... and the change is compatibile with virtually all packages. The idea is to have an additional value which always gets removed from objects. Thus allows you to construct your own instnces with 3state logic. I used `Possible` for that.

For me this is a highly usefull feature, e.g. when comunicating with APIs where one can request a subset of fields, or send update only for subset of them (and we have `null` as removing property), especially with GoogleApis (eg. https://github.com/tolysz/google-api/blob/master/src/Network/Google/Api/Types/GoogleUser.hs )

**Comment by phadej** [comment]
To repeat myself: I'm strongly against, :-1: adding `Omitted` or `Missing` to `Value`. There is no such constructor in JSON,

It's an encoding detail for narrow use case. Especially I dislike the idea of `decode . encode` being not `Just`, as it will lose information.

One can add handling of `Possible` to own generic implementation of record serialisation.

**Comment by ryantrinkle** [comment]
@phadej Agreed: it is extremely surprising if `decode . encode` is not `Just`.

**Comment by tolysz** [comment]
Generic or not I will not be able to use "toJSON", but I will be able to do it via toEncoding, thus it is a partial answer.
I.e. the way we serialize "Null" depends on some additional setting thus sometimes it is encoded as "" or "null",
And you can not possibly know how your instance will be encoded (all that is depending on setting is in use on the outside record) and this worries me ;)
And all that strong opposition to not encode this very information inside type itself (rather than in the function) is really surprising, and we only need one extra constructor (used only internally)

The specification does not say that some arbitrary key alway has to be removed or always has to be serialized with "null"... it does not say it can not be present for the same key for different cases, to possibly mean something different.
The specification only says how all possible strings have to look like after they are encoded, not how the internal structure should look like. 
Because of the `omittableNothing` one can not possibly know, by looking at the type how Maybe will behave, as one can set it to some arbitrary value - we need to know this setting.
Thus you need to know: `Null Bool`. I only propose to split this into 2 constructors, transparent and without that extra setting. 

> decode . encode is not Just. 

I am lost here Maybe will have exactly the same properties as it has now(i.e. settable via ommitNothing), the only difference will be: one could create newtypes AlwaysNullMaybe, AllwaysMissingMaybe with appropriate Value representation and rendering.

The Change I propose does not change anything in the current behavior, it simplifies things, it is not really visible, and one can be happily ignoring its existence...  Exactly like Imaginary part of complex numbers: the world description does not show imaginary numbers, to paraphrase the JSON RFC argument.


```
type Bla = Maybe Int
```
How will It be serialized?
If used on its own and if used inside a record. 

```
Foo = Foo { bla1 :: Bla, bla2 :: Bla}
Bar = Bar { bla1 :: Bla, bla2 :: Bla}
```
VS.
```
Foo = Foo { bla1:: AlwaysNullMaybe Int, bla2 :: AllwaysMissingMaybe Int}
Foo = Foo { bla1:: Possible Int, bla2:: Possible Int}
```

**Comment by mPiotrek** [comment]
I'll just add that it's not only a problem with nested `Maybe`s. It also breaks when using `Maybe Value`:

```haskell
fromJSON . toJSON $ Just Null :: Result (Maybe Value) = Success Nothing
```

**Comment by tolysz** [comment]
It will never be fixed... `2 != 3`... not until aeson discovers zero...
My stab at fixing it a while back:
https://github.com/haskell/aeson/compare/master...tolysz:aeson:Missing_1.5

**Comment by phadej** [comment]
Yes, intancer for `Maybe` won't change. It's used to encode missing values, and full roundtrip is not a priority.

**Comment by kindaro** [comment]
For whom is roundtrip not a priority? For me, roundtrip is a priority.

**Comment by cgibbard** [comment]
For what it's worth, round-tripping is very often the *only* property of the encoding we care about, apart from vague human readability for the sake of occasional debugging. Our clients and servers are both written in Haskell using aeson, and unless something goes wrong like this, we often don't even look at the wire format in detail, we're more likely to inspect the Haskell data structures on either end (apart from the case of looking at the communication in the web browser's network tool, which is sometimes useful).

**Comment by phadej** [comment]
The `Maybe` instance wont' change. It would be silent change which will break too much code out there without any notice from compiler. It just won't happen. maybe it wasn't a best choice 15 years or so ago, but now we have to live with it. I'm sorry.

**Comment by cgibbard** [comment]
Yeah, I wouldn't be in favour of a *silent* change to this behaviour in any case (though perhaps a major version or fork could eventually correct it). If we had a way to do something which generated a warning in cases where this was happening, that would be quite helpful.

**Comment by Lysxia** [comment]
The good thing about deriving implementations is that they don't need to live in the same package as their class. You can package your own generic/TH implementation of aeson's `ToJSON`/`FromJSON`. In fact there is already one out there: check out [deriving-aeson](https://github.com/fumieval/deriving-aeson). See if it has the same issue, and if yes, talk about it there.

aeson's generic deriving code is really not pleasant to maintain. It's best to consider it frozen legacy at this point.

**Comment by kindaro** [comment]
@phadej 

> I'm sorry.

I was furious at first, but I forgive you.

**Comment by kindaro** [comment]
@Lysxia   But this other package does not help if I want to serialize nested `Maybe`, right? Since there is no way to import the encoders and decoders from `aeson` without also importing the instances that are already defined, the broken `Maybe` instance will follow us everywhere, forever. So, this other package does not really solve the problem. The only way is to use another library for encoding and decoding.

**Comment by Lysxia** [comment]
Oh you're right. I don't know why I thought we were talking about `OmitNothingFields`

**Comment by phadej** [comment]
A core problem with "make all ToJSON/FromJSON instances roundtrip" is that it's near impossible virtually to enforce with parametric instances. E.g .currently

```haskell
instance ToJSON a => ToJSON (Maybe a) where
  toJSON Nothing  = Null
  toJSON (Just x) = toJSON x
```

We could (not really) treat this special case by introducing `UsesNull` as proposed early in the issue discussion i.e

```haskell
instance (ToJSON a, UsesNull a ~ False) => ToJSON (Maybe a) where
  UsesNull (Maybe a) = True
  ...
```

This has few drawbacks:
- It's a special case for Null
- It would require changing (and auditing) all the instances out there.
- ... because "correctness" of Maybe instance would depend on correctness of its parameter instance

Another option is to enforce injectivity structurally, e.g. `Either` instance does that

```haskell
instance ToJSON a => ToJSON (Either a b) where
  toJSON (Left y)  = object [ "Left" .= y ]
  toJSON (Right x) = object [ "Right" .= x ]
```

but changing `Maybe` to something like

```haskell
instance ToJSON a => ToJSON (Either a b) where
  toJSON Nothing  = "Nothing"
  toJSON (Just x) = object [ "Just".= x ]
```

doesn't feel right. (The `Either` instance feels off already).

It's actually natural to ask for `Either` instance to be just

```haskell
instance ToJSON a => ToJSON (Either a b) where
  toJSON (Left y)  = toJSON y
  toJSON (Right x) = toJSON x
```

and I have written such instances for specific sums where I know that parts dont overlap, e.g. something isomorphic to `Either Int String`. But generally it would make `Either` isntance quite dangerous as overlap risk is high, e.g. `Either Int Int` would lose the extra bit of information, not clear what `Either Int Double` does etc.

I should emphasise that

```haskell
instance ToJSON a => ToJSON (Maybe a) where
  toJSON Nothing  = Null
  toJSON (Just x) = toJSON x
```

instance is very natural one. Using it with types which themselves don't use `null` in their encoding (e.g. all the objects etc.) is great. So changing the "shape" of this instance is not an option.

So there are only bad and worse options on the table
- Add UsesNull - requires everyone to change (as default should be conservative, i.e. `True`), and it's a quite specific solution, just for `Maybe` instance.  Also error prone.
- Change `Maybe` instance to be injective by construction: huge change.
- Keep things they are. This is "bad" option too, but less worse then others.

If there were a way to somehow "refine" encodings, so `Maybe` instance could require for encoding to not be `null` or future  `Either` instance to ask for parts to be disjoint, that would be great. `aeson` could adopt it if it's sufficiently lightweight and easy to use.

Until then, there is no point in discussing problems with `Maybe (Maybe a)`. If it's your problem, don't use `Maybe (Maybe a)` define your own type to handle that if you need injectivity of encoding.

Lastly, I don't like that we have special handling of `String` (i.e. `[Char]`) in `aeson`. It adds incidental complexity everywhee.  Additionally handling `Maybe` specially in the `ToJSON/FromJSON` instance would make instance machinery even worse.

---

The recent discussion didn't add anything to the discussion from either years ago, so @kindaro until you propose something actually with a reasonably potential, the `Maybe` instance we have is the best we can have at the moment.

---

My bad that I didn't close this issue sooner (e.g. four years ago when it was last brought to my attention).

**Comment by tolysz** [comment]
` Maybe (Maybe a)` ~ 3 values: Value, null, removed key... useful for things like Google Api, or Mongo where... setting something to null removes the value, and not including the key keeps the current value.

So when you would extend your type on the server you would not force people to rewrite/rebuild your clients.

**Comment by phadej** [comment]
@tolysz "not inclduing the key" is not something `ToJSON` instance can affect. It could be an option in `ToJSONField` class used to compose object encodings, but `aeson` doesn't have such.

your comment is off-topic.

**Comment by kindaro** [comment]
I understand that there is a fundamental, cursed conflict here. We cannot at the same time have:

* mathematical correctness
* convenience for _«lowest common denominator»_ computer programmers

My error was thinking that Haskell is a programming language where, culturally, mathematical correctness is valued above convenience for _«lowest common denominator»_ computer programmers. I was misled to thinking this way by the _«ambient»_ marketing where things like _«functor laws»_ and such are, seemingly, promoted. Based on a number of observations, including this issue, I accept that I was wrong in thinking this way.

In the case of encoding, serialization and such, the law of inversion `decode ∘ encode = identity` is the obvious one to respect, and `aeson` does not respect it.

It is not necessarily wrong in principle to put convenience for _«lowest common denominator»_ computer programmers above mathematical correctness. It is simply not what I want from my libraries. So this library is simply not for me.
